File: KNIRVENGINE/desktop-client/tests/knirvoracle_integration_test.go
package tests

import (
	"context"
	"os"
	"testing"
	"time"

	"KNIRVENGINE/desktop-client/agent"
	"KNIRVENGINE/desktop-client/services"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// TestKNIRVOracleIntegration tests the complete KNIRVORACLE integration
func TestKNIRVOracleIntegration(t *testing.T) {
	// Skip if KNIRVORACLE_URL is not set (for CI/CD environments)
	knirvoracleURL := os.Getenv("KNIRVORACLE_URL")
	if knirvoracleURL == "" {
		t.Skip("KNIRVORACLE_URL not set, skipping integration tests")
	}

	// Initialize KNIRVORACLE service
	config := services.KNIRVOracleConfig{
		BaseURL: knirvoracleURL,
		APIKey:  os.Getenv("KNIRVORACLE_API_KEY"),
		Timeout: 30,
	}
	knirvoracleService := services.NewKNIRVOracleService(config)

	t.Run("HealthCheck", func(t *testing.T) {
		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
		defer cancel()

		err := knirvoracleService.HealthCheck(ctx)
		assert.NoError(t, err, "KNIRVORACLE health check should pass")
	})

	t.Run("AgentMinting", func(t *testing.T) {
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		// Create test agent configuration
		agentConfig := agent.AgentConfig{
			AgentID:     "test-agent-" + time.Now().Format("20060102150405"),
			Name:        "Test Agent",
			Description: "Test agent for integration testing",
			AgentType:   "assistant",
			Model:       "gpt-4",
		}

		// Test agent minting
		mintRequest := services.AgentMintRequest{
			AgentID:     agentConfig.AgentID,
			Name:        agentConfig.Name,
			Description: agentConfig.Description,
			Owner:       "test-owner",
			Metadata: map[string]interface{}{
				"agent_type": agentConfig.AgentType,
				"model":      agentConfig.Model,
				"test":       true,
			},
		}

		response, err := knirvoracleService.MintAgent(ctx, mintRequest)
		require.NoError(t, err, "Agent minting should not fail")
		assert.True(t, response.Success, "Agent minting should succeed")
		assert.NotEmpty(t, response.TransactionID, "Transaction ID should be provided")
		assert.NotEmpty(t, response.AgentNFTID, "Agent NFT ID should be provided")
	})

	t.Run("CapabilityRegistration", func(t *testing.T) {
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		// Test capability registration
		capabilityRequest := services.CapabilityRegistrationRequest{
			Name:        "Test Capability " + time.Now().Format("15:04:05"),
			Type:        "mcp_capability",
			Description: "Test capability for integration testing",
			Schema: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"input": map[string]interface{}{
						"type":        "string",
						"description": "Input data for the capability",
					},
				},
				"required": []string{"input"},
			},
			Owner:     "test-owner",
			GasFeeNRN: 1000,
		}

		response, err := knirvoracleService.RegisterCapability(ctx, capabilityRequest)
		require.NoError(t, err, "Capability registration should not fail")
		assert.True(t, response.Success, "Capability registration should succeed")
		assert.NotEmpty(t, response.CapabilityID, "Capability ID should be provided")
		assert.NotEmpty(t, response.TxHash, "Transaction hash should be provided")
	})

	t.Run("WalletOperations", func(t *testing.T) {
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		// Test wallet address (use a test address)
		testAddress := "0x742d35Cc6aa34567890abcdef1234567890abcdef"

		// Test faucet request
		faucetRequest := services.FaucetRequest{
			Address: testAddress,
			Amount:  "100",
			Reason:  "Integration test",
		}

		faucetResponse, err := knirvoracleService.RequestFaucet(ctx, faucetRequest)
		if err != nil {
			t.Logf("Faucet request failed (may be expected): %v", err)
		} else {
			assert.True(t, faucetResponse.Success, "Faucet request should succeed")
			assert.NotEmpty(t, faucetResponse.RequestID, "Request ID should be provided")
		}

		// Test wallet balance retrieval
		balanceResponse, err := knirvoracleService.GetWalletBalance(ctx, testAddress)
		if err != nil {
			t.Logf("Balance retrieval failed (may be expected for test address): %v", err)
		} else {
			assert.True(t, balanceResponse.Success, "Balance retrieval should succeed")
			assert.Equal(t, testAddress, balanceResponse.Address, "Address should match")
		}
	})

	t.Run("TreasuryStatus", func(t *testing.T) {
		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
		defer cancel()

		treasuryStatus, err := knirvoracleService.GetTreasuryStatus(ctx)
		if err != nil {
			t.Logf("Treasury status retrieval failed (may be expected): %v", err)
		} else {
			assert.NotNil(t, treasuryStatus, "Treasury status should be returned")
		}
	})
}

// TestAgentBuilderKNIRVOracleIntegration tests the agent builder integration with KNIRVORACLE
func TestAgentBuilderKNIRVOracleIntegration(t *testing.T) {
	// Skip if KNIRVORACLE_URL is not set
	if os.Getenv("KNIRVORACLE_URL") == "" {
		t.Skip("KNIRVORACLE_URL not set, skipping integration tests")
	}

	// Create temporary directory for test
	tempDir := t.TempDir()

	// Initialize agent builder
	builder, err := agent.NewAgentBuilderWithStorage(nil, "../../agent/templates", tempDir)
	require.NoError(t, err, "Agent builder initialization should not fail")

	t.Run("BuildAgentWithMinting", func(t *testing.T) {
		// Create test agent configuration
		agentConfig := agent.AgentConfig{
			Name:        "Integration Test Agent",
			Description: "Agent created during integration testing",
			AgentType:   "assistant",
			Model:       "gpt-4",
			ExtraParams: map[string]interface{}{
				"test": true,
			},
		}

		// Build the agent (this should trigger minting)
		agentID, err := builder.BuildAgent(agentConfig)
		require.NoError(t, err, "Agent building should not fail")
		assert.NotEmpty(t, agentID, "Agent ID should be provided")

		// Wait a bit for the minting process to complete
		time.Sleep(5 * time.Second)

		// Verify the agent was created
		retrievedAgent, err := builder.GetAgent(agentID)
		require.NoError(t, err, "Agent retrieval should not fail")
		assert.Equal(t, agentConfig.Name, retrievedAgent.Name, "Agent name should match")
		assert.Equal(t, agentConfig.Description, retrievedAgent.Description, "Agent description should match")
	})
}

// TestKNIRVOracleServiceConfiguration tests service configuration
func TestKNIRVOracleServiceConfiguration(t *testing.T) {
	t.Run("DefaultConfiguration", func(t *testing.T) {
		config := services.KNIRVOracleConfig{
			BaseURL: "http://localhost:8080",
			Timeout: 30,
		}

		service := services.NewKNIRVOracleService(config)
		assert.NotNil(t, service, "Service should be created")
	})

	t.Run("ConfigurationWithAPIKey", func(t *testing.T) {
		config := services.KNIRVOracleConfig{
			BaseURL: "http://localhost:8080",
			APIKey:  "test-api-key",
			Timeout: 60,
		}

		service := services.NewKNIRVOracleService(config)
		assert.NotNil(t, service, "Service should be created with API key")
	})
}

// TestKNIRVOracleErrorHandling tests error handling scenarios
func TestKNIRVOracleErrorHandling(t *testing.T) {
	// Create service with invalid URL
	config := services.KNIRVOracleConfig{
		BaseURL: "http://invalid-url:9999",
		Timeout: 5,
	}
	service := services.NewKNIRVOracleService(config)

	t.Run("HealthCheckFailure", func(t *testing.T) {
		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		defer cancel()

		err := service.HealthCheck(ctx)
		assert.Error(t, err, "Health check should fail with invalid URL")
	})

	t.Run("AgentMintingFailure", func(t *testing.T) {
		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		defer cancel()

		mintRequest := services.AgentMintRequest{
			AgentID:     "test-agent",
			Name:        "Test Agent",
			Description: "Test agent",
			Owner:       "test-owner",
			Metadata:    map[string]interface{}{},
		}

		_, err := service.MintAgent(ctx, mintRequest)
		assert.Error(t, err, "Agent minting should fail with invalid URL")
	})
}

// BenchmarkKNIRVOracleOperations benchmarks KNIRVORACLE operations
func BenchmarkKNIRVOracleOperations(b *testing.B) {
	// Skip if KNIRVORACLE_URL is not set
	if os.Getenv("KNIRVORACLE_URL") == "" {
		b.Skip("KNIRVORACLE_URL not set, skipping benchmarks")
	}

	config := services.KNIRVOracleConfig{
		BaseURL: os.Getenv("KNIRVORACLE_URL"),
		APIKey:  os.Getenv("KNIRVORACLE_API_KEY"),
		Timeout: 30,
	}
	service := services.NewKNIRVOracleService(config)

	b.Run("HealthCheck", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			service.HealthCheck(ctx)
			cancel()
		}
	})

	b.Run("GetTreasuryStatus", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			service.GetTreasuryStatus(ctx)
			cancel()
		}
	})
}

File: packages/KNIRVENGINE/gui/src/tests/knirvoracleIntegration.test.ts
import { KNIRVOracleService } from '../services/knirvoracleService';
import { WalletService } from '../services/walletService';

type MockClient = {
  post: jest.Mock;
  get: jest.Mock;
};

const getClient = (service: KNIRVOracleService): MockClient =>
  (service as unknown as { client: MockClient }).client;

describe('KNIRVORACLE Integration Tests', () => {
  let knirvoracleService: KNIRVOracleService;
  let walletService: WalletService;

  beforeEach(() => {
    // Initialize services with test configuration
    knirvoracleService = new KNIRVOracleService({
      baseURL: import.meta.env.VITE_KNIRVORACLE_URL || 'http://localhost:8080',
      apiKey: import.meta.env.VITE_KNIRVORACLE_API_KEY || 'test-key',
      timeout: 10000,
    });

    walletService = new WalletService();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('KNIRVOracleService', () => {
    it('should initialize with correct configuration', () => {
      expect(knirvoracleService).toBeDefined();
      expect(knirvoracleService.getConfig().baseURL).toBe(
        import.meta.env.VITE_KNIRVORACLE_URL || 'http://localhost:8080'
      );
    });

    it('should update configuration correctly', () => {
      const newConfig = {
        baseURL: 'http://new-url:8080',
        apiKey: 'new-api-key',
      };

      knirvoracleService.updateConfig(newConfig);
      const config = knirvoracleService.getConfig();

      expect(config.baseURL).toBe(newConfig.baseURL);
      expect(config.apiKey).toBe(newConfig.apiKey);
    });

    it('should handle agent minting request', async () => {
      const mockResponse = {
        success: true,
        transaction_id: 'tx-123',
        agent_nft_id: 'nft-456',
        message: 'Agent minted successfully',
      };

      // Mock the axios post method
      const mockPost = jest.fn().mockResolvedValue({ data: mockResponse });
      getClient(knirvoracleService).post = mockPost;

      const mintRequest = {
        agent_id: 'test-agent-123',
        name: 'Test Agent',
        description: 'A test agent for integration testing',
        owner: 'test-owner',
        metadata: {
          type: 'assistant',
          model: 'gpt-4',
        },
      };

      const result = await knirvoracleService.mintAgent(mintRequest);

      expect(mockPost).toHaveBeenCalledWith('/agent/mint', mintRequest, { timeout: 60000 });
      expect(result).toEqual(mockResponse);
      expect(result.success).toBe(true);
      expect(result.transaction_id).toBe('tx-123');
      expect(result.agent_nft_id).toBe('nft-456');
    });

    it('should handle capability registration', async () => {
      const mockResponse = {
        success: true,
        capability_id: 'cap-789',
        tx_hash: '0xabcdef123456',
        message: 'Capability registered successfully',
      };

      const mockPost = jest.fn().mockResolvedValue({ data: mockResponse });
      getClient(knirvoracleService).post = mockPost;

      const registrationRequest = {
        name: 'Test Capability',
        type: 'mcp_capability',
        description: 'A test capability',
        schema: {
          type: 'object',
          properties: {
            input: { type: 'string' },
          },
        } as const,
        owner: 'test-owner',
        gas_fee_nrn: 1000,
      };

      const result = await knirvoracleService.registerCapability(registrationRequest);

      expect(mockPost).toHaveBeenCalledWith('/wallet/mcp/create_register_capability', registrationRequest, { timeout: 30000 });
      expect(result).toEqual(mockResponse);
      expect(result.success).toBe(true);
      expect(result.capability_id).toBe('cap-789');
    });

    it('should handle faucet requests', async () => {
      const mockResponse = {
        success: true,
        request_id: 'req-123',
        tx_hash: '0x123456789',
        amount: '100',
        status: 'completed',
        message: 'Faucet request successful',
      };

      const mockPost = jest.fn().mockResolvedValue({ data: mockResponse });
      getClient(knirvoracleService).post = mockPost;

      const faucetRequest = {
        address: '0x742d35Cc6aa34567890abcdef1234567890abcdef',
        amount: '100',
        reason: 'Test faucet request',
      };

      const result = await knirvoracleService.requestFaucet(faucetRequest);

      expect(mockPost).toHaveBeenCalledWith('/api/mint/nrv', faucetRequest);
      expect(result).toEqual(mockResponse);
      expect(result.success).toBe(true);
      expect(result.amount).toBe('100');
    });

    it('should handle wallet balance queries', async () => {
      const mockResponse = {
        success: true,
        address: '0x742d35Cc6aa34567890abcdef1234567890abcdef',
        balance: '1000.50',
        nrn_balance: '1247',
        usd_value: '312.75',
      };

      const mockGet = jest.fn().mockResolvedValue({ data: mockResponse });
      getClient(knirvoracleService).get = mockGet;

      const address = '0x742d35Cc6aa34567890abcdef1234567890abcdef';
      const result = await knirvoracleService.getWalletBalance(address);

      expect(mockGet).toHaveBeenCalledWith(`/balance/${address}`);
      expect(result).toEqual(mockResponse);
      expect(result.success).toBe(true);
      expect(result.address).toBe(address);
    });

    it('should handle transaction sending', async () => {
      const mockResponse = {
        success: true,
        transaction_id: 'tx-789',
        tx_hash: '0xdef456789',
        status: 'pending',
        message: 'Transaction submitted',
      };

      const mockPost = jest.fn().mockResolvedValue({ data: mockResponse });
      getClient(knirvoracleService).post = mockPost;

      const transactionRequest = {
        from: '0x742d35Cc6aa34567890abcdef1234567890abcdef',
        to: '0x987654321fedcba0987654321fedcba0987654321',
        amount: '50',
        token: 'NRN',
        memo: 'Test transaction',
      };

      const result = await knirvoracleService.sendTransaction(transactionRequest);

      expect(mockPost).toHaveBeenCalledWith('/transactions', transactionRequest);
      expect(result).toEqual(mockResponse);
      expect(result.success).toBe(true);
      expect(result.transaction_id).toBe('tx-789');
    });

    it('should handle health checks', async () => {
      const mockGet = jest.fn().mockResolvedValue({ status: 200 });
      getClient(knirvoracleService).get = mockGet;

      const result = await knirvoracleService.healthCheck();

      expect(mockGet).toHaveBeenCalledWith('/health', { timeout: 5000 });
      expect(result).toBe(true);
    });

    it('should handle health check failures', async () => {
      const mockGet = jest.fn().mockRejectedValue(new Error('Network error'));
      getClient(knirvoracleService).get = mockGet;

      const result = await knirvoracleService.healthCheck();

      expect(result).toBe(false);
    });

    it('should handle capability invocation', async () => {
      const mockResponse = {
        success: true,
        result: 'Capability executed successfully',
        execution_time: 1500,
      };

      const mockPost = jest.fn().mockResolvedValue({ data: mockResponse });
      getClient(knirvoracleService).post = mockPost;

      const capabilityId = 'cap-123';
      const inputData = { input: 'test data' };

      const result = await knirvoracleService.invokeCapability(capabilityId, inputData);

      expect(mockPost).toHaveBeenCalledWith('/wallet/mcp/create_invoke_capability', {
        capability_id: capabilityId,
        interaction_type: 'invoke',
        input_data: inputData,
        timestamp: expect.any(Number),
      });
      expect(result).toEqual(mockResponse);
    });

    it('should handle error responses gracefully', async () => {
      const mockPost = jest.fn().mockRejectedValue(new Error('API Error'));
      getClient(knirvoracleService).post = mockPost;

      const mintRequest = {
        agent_id: 'test-agent',
        name: 'Test Agent',
        description: 'Test',
        owner: 'test-owner',
        metadata: {},
      };

      await expect(knirvoracleService.mintAgent(mintRequest)).rejects.toThrow('Failed to mint agent');
    });
  });

  describe('Wallet Service KNIRVORACLE Integration', () => {
    it('should integrate with KNIRVORACLE for balance queries', async () => {
      // This test would verify that the wallet service correctly calls KNIRVORACLE
      // In a real scenario, we would mock the KNIRVORACLE service calls
      expect(walletService).toBeDefined();
    });

    it('should fallback to local API when KNIRVORACLE fails', async () => {
      // This test would verify fallback behavior
      expect(walletService).toBeDefined();
    });
  });

  describe('Type Safety', () => {
    it('should enforce correct types for agent minting', () => {
      const mintRequest = {
        agent_id: 'test-agent',
        name: 'Test Agent',
        description: 'Test description',
        owner: 'test-owner',
        metadata: {
          type: 'assistant',
          model: 'gpt-4',
        },
      };

      // TypeScript should enforce these types at compile time
      expect(typeof mintRequest.agent_id).toBe('string');
      expect(typeof mintRequest.name).toBe('string');
      expect(typeof mintRequest.metadata).toBe('object');
    });

    it('should enforce correct types for capability registration', () => {
      const registrationRequest = {
        name: 'Test Capability',
        type: 'mcp_capability',
        description: 'Test description',
        schema: {
          type: 'object',
          properties: {},
        },
        owner: 'test-owner',
        gas_fee_nrn: 1000,
      };

      expect(typeof registrationRequest.name).toBe('string');
      expect(typeof registrationRequest.gas_fee_nrn).toBe('number');
      expect(typeof registrationRequest.schema).toBe('object');
    });
  });

  describe('Error Handling', () => {
    it('should handle network timeouts', async () => {
      const mockPost = jest.fn().mockRejectedValue(new Error('timeout'));
      getClient(knirvoracleService).post = mockPost;

      const mintRequest = {
        agent_id: 'test-agent',
        name: 'Test Agent',
        description: 'Test',
        owner: 'test-owner',
        metadata: {},
      };

      await expect(knirvoracleService.mintAgent(mintRequest)).rejects.toThrow();
    });

    it('should handle invalid responses', async () => {
      const mockPost = jest.fn().mockResolvedValue({ data: null });
      getClient(knirvoracleService).post = mockPost;

      const mintRequest = {
        agent_id: 'test-agent',
        name: 'Test Agent',
        description: 'Test',
        owner: 'test-owner',
        metadata: {},
      };

      const result = await knirvoracleService.mintAgent(mintRequest);
      expect(result).toBeNull();
    });
  });
});
